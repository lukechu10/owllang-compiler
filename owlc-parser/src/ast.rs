//! Syntax representation for owllang source code.
//! Includes all the abstract syntax tree nodes used in other crates.

pub mod expressions;
pub mod statements;

