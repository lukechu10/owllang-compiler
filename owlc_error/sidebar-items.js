initSidebarItems({"struct":[["Error","Represents a compile error."],["ErrorReporter","Handles all errors for a compilation unit. Use `format!(\"{}\", err_reporter);` to retrieve list of errors."]]});