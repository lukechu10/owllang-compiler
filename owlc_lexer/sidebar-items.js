initSidebarItems({"enum":[["TokenKind","Represents the value of a [`Token`]."]],"struct":[["Lexer",""],["Token","Represents a single token."]]});