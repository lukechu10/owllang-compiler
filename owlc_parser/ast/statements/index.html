<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `statements` mod in crate `owlc_parser`."><meta name="keywords" content="rust, rustlang, rust-lang, statements"><title>owlc_parser::ast::statements - Rust</title><link rel="stylesheet" type="text/css" href="../../../normalize.css"><link rel="stylesheet" type="text/css" href="../../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../../ayu.css" disabled ><script src="../../../storage.js"></script><noscript><link rel="stylesheet" href="../../../noscript.css"></noscript><link rel="shortcut icon" href="../../../favicon.ico"><style type="text/css">#crate-search{background-image:url("../../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../../../owlc_parser/index.html'><div class='logo-container rust-logo'><img src='../../../rust-logo.png' alt='logo'></div></a><p class='location'>Module statements</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li></ul></div><p class='location'><a href='../../index.html'>owlc_parser</a>::<wbr><a href='../index.html'>ast</a></p><script>window.sidebarCurrent = {name: 'statements', ty: 'mod', relpath: '../'};</script><script defer src="../sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!"><img src="../../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices"></div></div><script src="../../../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><span class="help-button">?</span>
                <a id="settings-menu" href="../../../settings.html"><img src="../../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class='fqn'><span class='out-of-band'><span id='render-detail'><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class='inner'>&#x2212;</span>]</a></span><a class='srclink' href='../../../src/owlc_parser/ast/statements.rs.html#1-121' title='goto source code'>[src]</a></span><span class='in-band'>Module <a href='../../index.html'>owlc_parser</a>::<wbr><a href='../index.html'>ast</a>::<wbr><a class="mod" href=''>statements</a></span></h1><h2 id='structs' class='section-header'><a href="#structs">Structs</a></h2>
<table><tr class='module-item'><td><a class="struct" href="struct.Block.html" title='owlc_parser::ast::statements::Block struct'>Block</a></td><td class='docblock-short'><p>Represents a block. e.g. <code>{...}</code>. A <code>Block</code> creates a new lexical scope. Variables declared inside the <code>Block</code> are not visible outside the <code>Block</code>.</p>
</td></tr><tr class='module-item'><td><a class="struct" href="struct.CompilationUnit.html" title='owlc_parser::ast::statements::CompilationUnit struct'>CompilationUnit</a></td><td class='docblock-short'><p>Represents a collection of functions that are included in the compilation unit.</p>
</td></tr><tr class='module-item'><td><a class="struct" href="struct.FnProto.html" title='owlc_parser::ast::statements::FnProto struct'>FnProto</a></td><td class='docblock-short'><p>Represents a function prototype (not the function body). e.g. <code>fn f(x)</code>.</p>
</td></tr><tr class='module-item'><td><a class="struct" href="struct.Stmt.html" title='owlc_parser::ast::statements::Stmt struct'>Stmt</a></td><td class='docblock-short'><p>Represents a statement. Statements do not have any value unlike <a href="../expressions/struct.Expr.html"><code>Expr</code></a>. Statements usually have side effects.</p>
</td></tr></table><h2 id='enums' class='section-header'><a href="#enums">Enums</a></h2>
<table><tr class='module-item'><td><a class="enum" href="enum.StmtKind.html" title='owlc_parser::ast::statements::StmtKind enum'>StmtKind</a></td><td class='docblock-short'><p>Internal representation for <a href="struct.Stmt.html"><code>Stmt</code></a>.</p>
</td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../../../";window.currentCrate = "owlc_parser";</script><script src="../../../main.js"></script><script defer src="../../../search-index.js"></script></body></html>