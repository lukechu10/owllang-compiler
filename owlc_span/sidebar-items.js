initSidebarItems({"struct":[["BytePos","Utility for storing position in source code. Position is represented as the position of the byte in the `SourceFile`. `BytePos` should be 0 based."],["SourceFile","Represents 1 `SourceFile`. NOTE: A `CompilationUnit` can have multiple `SourceFile`s."],["Span","`Span` represents a region of code, used for error reporting. `Span` is represented with the starting and ending position of the region. The starting position is inclusive and the ending position is exclusive."]],"type":[["FileName","Use `Rc` to prevent unnecessary duplication of the `String`."]]});