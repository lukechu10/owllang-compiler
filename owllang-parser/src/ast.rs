pub mod expressions;
pub mod statements;