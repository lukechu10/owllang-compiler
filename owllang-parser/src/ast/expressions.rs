use owllang_lexer::TokenKind;

#[derive(Debug)]
pub enum ExprKind {
    /// Represents an int literal (internally represented using `i64`).
    Literal(i64),
    /// Represents an identifier expression.
    Identifier(String),
    /// Represents a function call expression.
    FuncCall { callee: String, args: Vec<Expr> },
    BinaryExpr {
        lhs: Box<Expr>,
        rhs: Box<Expr>,
        /// The operator type of the `BinaryExpr`. Represented with a `TokenVal`. The value of the field should only be valid operator variants of `TokenVal`.
        op_type: TokenKind,
    },
}

#[derive(Debug)]
/// Represents an expression.
pub struct Expr {
    pub kind: ExprKind,
}
impl Expr {
    pub fn new(kind: ExprKind) -> Self {
        Self { kind }
    }
}
