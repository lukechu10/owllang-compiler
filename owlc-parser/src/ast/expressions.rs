use owlc_lexer::TokenKind;
use owlc_span::Span;
use serde::Serialize;

/// Internal representation for [`Expr`](struct.Expr.html).
#[derive(Serialize, Debug)]
pub enum ExprKind {
    /// Represents an int literal (internally represented using `i64`).
    Literal(i64),
    /// Represents an identifier expression.
    Identifier(String),
    /// Represents a function call expression.
    FuncCall { callee: String, args: Vec<Expr> },
    BinaryExpr {
        lhs: Box<Expr>,
        rhs: Box<Expr>,
        /// The operator type of the `BinaryExpr`. Represented with a `TokenVal`. The value of the field should only be valid operator variants of `TokenVal`.
        op_type: TokenKind,
    },
}

/// Represents an expression. Expressions may or may not have a side effect.
/// # Example
/// ```owllang
/// 1 + 1 // no side effect here
/// println(3) // side effect (printing to stdout)
/// ```
#[derive(Serialize, Debug)]
pub struct Expr {
    #[serde(flatten)]
    pub kind: ExprKind,
    #[serde(skip_serializing)]
    pub span: Span,
}
