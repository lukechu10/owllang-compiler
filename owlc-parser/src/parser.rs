//! Parsing logic. Transforms `Token`s into abstract syntax tree nodes.

use crate::ast::{expressions::*, statements::*};
use owlc_error::{Error, ErrorReporter};
use owlc_lexer::{Lexer, Token, TokenKind};
use owlc_span::{BytePos, SourceFile};
use std::iter::Peekable;
use std::rc::Rc;

#[derive(Debug)]
pub struct Parser<'a> {
    src: Rc<SourceFile>,
    lexer: Peekable<&'a mut Lexer<'a>>,
    current_token: Token,

    errs: &'a mut ErrorReporter,
}

impl<'a> Parser<'a> {
    pub fn new(lexer: &'a mut Lexer<'a>, error_reporter: &'a mut ErrorReporter) -> Self {
        let first_token = lexer.next().unwrap_or(Token {
            kind: TokenKind::EndOfFile,
            span: BytePos(0).to(BytePos(1)),
        });

        Self {
            src: Rc::clone(&lexer.src),
            lexer: lexer.peekable(),
            current_token: first_token,
            errs: error_reporter,
        }
    }

    /// Creates a new syntax error at the current token and adds it to the `ErrorReporter`.
    fn emit_err_at_current_tok(&mut self, message: String) {
        let err = Error::new(self.src.name.to_string(), self.current_token.span, message);

        self.errs.report(err);
    }

    /// Returns the token that was eaten. If `self.lexer` has no more `Token`s left, returns an `TokenKind::EndOfFile`.
    fn eat_token(&mut self) -> Token {
        let tmp_token = self.current_token.clone();
        match self.lexer.next() {
            Some(tok) => self.current_token = tok,
            None => {
                self.current_token = Token {
                    kind: TokenKind::EndOfFile,
                    span: self.current_token.span.hi.zero_width_span(),
                }
            }
        }
        tmp_token
    }

    /// Emits an error and constructs expected token even if bad match.
    fn expect_and_eat_tok(&mut self, expected: TokenKind) -> Token {
        let actual = self.eat_token();
        if actual.kind != expected {
            // emit error
            self.emit_err_at_current_tok(format!(
                "Expected {} token but found {} token.",
                expected, actual.kind
            ));

            Token {
                kind: expected,
                span: self.current_token.span,
            }
        } else {
            actual
        }
    }

    /// Returns the identifier `String` or `"0_err_iden"` value if current token is not an identifier.
    /// Error identifier starts with `'0'` to prevent conflicts with valid identifiers.
    fn expect_and_eat_iden_tok(&mut self) -> String {
        let actual = self.eat_token();
        if let TokenKind::Identifier(iden) = actual.kind {
            iden
        } else {
            self.emit_err_at_current_tok(format!(
                "Expected a {} token but found {} token.",
                TokenKind::Identifier("".to_string()),
                actual.kind
            ));
            "0_err_iden".to_string()
        }
    }

    pub fn record_current_pos(&self) -> BytePos {
        self.current_token.span.lo
    }
}

impl<'a> Parser<'a> {
    /// User can input both fn definitions and statements / expressions in the repl prompt.
    pub fn parse_repl_input(&mut self) -> Stmt {
        match self.current_token.kind {
            TokenKind::EndOfFile => Stmt {
                kind: StmtKind::Noop,
                span: self.record_current_pos().zero_width_span(),
            },
            TokenKind::KeywordFn | TokenKind::KeywordExtern => self.parse_fn_declaration(),
            // TokenKind::KeywordLet => {
            //     let let_statement = self.parse_let_statement();
            //     if self.current_token.kind == TokenKind::PuncSemi {
            //         self.expect_and_eat_tok(TokenKind::PuncSemi);
            //     }
            //     let_statement
            // }
            _ => {
                // try to parse expression statement
                let expr = self.parse_expression();
                // semi colon is optional in repl
                if self.current_token.kind == TokenKind::PuncSemi {
                    self.expect_and_eat_tok(TokenKind::PuncSemi);
                }
                Stmt {
                    span: expr.span.lo.to(self.record_current_pos()),
                    kind: StmtKind::ExprSemi { expr },
                }
            }
        }
    }

    pub fn parse_compilation_unit(&mut self) -> CompilationUnit {
        let mut compilation_unit = CompilationUnit::new(self.src.name.to_string());
        while self.current_token.kind != TokenKind::EndOfFile {
            let fn_declaration = self.parse_fn_declaration();
            compilation_unit.add_func(fn_declaration);
        }
        compilation_unit
    }

    fn parse_statement(&mut self) -> Stmt {
        match self.current_token.kind {
            TokenKind::KeywordReturn => {
                let ret_statement = self.parse_return_statement();
                self.expect_and_eat_tok(TokenKind::PuncSemi);
                ret_statement
            }
            TokenKind::KeywordLet => {
                let let_statement = self.parse_let_statement();
                self.expect_and_eat_tok(TokenKind::PuncSemi);
                let_statement
            }
            TokenKind::KeywordWhile => self.parse_while_stmt(),
            TokenKind::KeywordIf => self.parse_if_else_stmt(),
            TokenKind::PuncOpenBrace => {
                let lo = self.record_current_pos();
                let block = self.parse_block_statement();
                let hi = self.record_current_pos();
                Stmt {
                    kind: StmtKind::Block(block),
                    span: lo.to(hi),
                }
            }
            _ => {
                // try to parse expression statement
                let expr = self.parse_expression();
                self.expect_and_eat_tok(TokenKind::PuncSemi);
                Stmt {
                    span: expr.span.lo.to(self.record_current_pos()),
                    kind: StmtKind::ExprSemi { expr },
                }
            }
        }
    }

    fn parse_return_statement(&mut self) -> Stmt {
        let lo = self.record_current_pos();
        self.expect_and_eat_tok(TokenKind::KeywordReturn);
        let value = self.parse_expression();
        let hi = self.record_current_pos();
        Stmt {
            kind: StmtKind::Return { value },
            span: lo.to(hi),
        }
    }

    fn parse_let_statement(&mut self) -> Stmt {
        let lo = self.record_current_pos();
        self.expect_and_eat_tok(TokenKind::KeywordLet);
        let iden = self.expect_and_eat_iden_tok();
        self.expect_and_eat_tok(TokenKind::OpEquals);
        let initializer = self.parse_expression();
        let hi = self.record_current_pos();
        Stmt {
            kind: StmtKind::Let { iden, initializer },
            span: lo.to(hi),
        }
    }

    /// Parses any valid expression.
    fn parse_expression(&mut self) -> Expr {
        // 0 binding power to accept any expression
        self.parse_binary_expr_rhs(0)
    }

    /// Parses an atomic expression. Returns a `'0'` literal expression if bad match.
    fn parse_primary_expression(&mut self) -> Expr {
        match self.current_token.kind {
            TokenKind::PuncOpenParen => {
                self.expect_and_eat_tok(TokenKind::PuncOpenParen);
                // parse expression inside parenthesis.
                let expr_ast = self.parse_expression();
                self.expect_and_eat_tok(TokenKind::PuncCloseParen);
                expr_ast
            }
            TokenKind::LiteralInt(_) => self.parse_int_literal(),
            TokenKind::Identifier(_) => self.parse_identifier_or_call_expr(),
            _ => {
                self.emit_err_at_current_tok(format!(
                    "Expected an expression but found {} token.",
                    self.current_token.kind
                ));
                Expr {
                    kind: ExprKind::Literal(0),
                    span: self.record_current_pos().zero_width_span(),
                }
            }
        }
    }

    /// Returns the parsed `IntLiteral` token or `0` if bad match.
    fn parse_int_literal(&mut self) -> Expr {
        let lo = self.record_current_pos();
        if let TokenKind::LiteralInt(num) = self.current_token.kind {
            self.eat_token(); // eat int literal token
            let hi = self.record_current_pos();
            Expr {
                kind: ExprKind::Literal(num),
                span: lo.to(hi),
            }
        } else {
            self.emit_err_at_current_tok(format!(
                "Expected {} token but found {} token.",
                self.current_token.kind,
                TokenKind::LiteralInt(0) // dummy value for generating message
            ));
            Expr {
                kind: ExprKind::Literal(0),
                span: lo.zero_width_span(),
            }
        }
    }

    /// Returns a `IdentifierExprAST` or `CallExprAST`, depending on the scenario.
    fn parse_identifier_or_call_expr(&mut self) -> Expr {
        let lo = self.record_current_pos();
        let identifier = self.expect_and_eat_iden_tok();

        if self.current_token.kind == TokenKind::PuncOpenParen {
            self.expect_and_eat_tok(TokenKind::PuncOpenParen);
            // parse function call expression
            let mut args: Vec<Expr> = Vec::new();
            loop {
                if self.current_token.kind == TokenKind::PuncCloseParen {
                    // found end of argument list
                    self.expect_and_eat_tok(TokenKind::PuncCloseParen);
                    break;
                }

                let expr = self.parse_expression();
                args.push(expr);

                if self.current_token.kind == TokenKind::PuncComma {
                    self.expect_and_eat_tok(TokenKind::PuncComma);
                } else if self.current_token.kind != TokenKind::PuncCloseParen {
                    self.emit_err_at_current_tok(format!(
                        "Expected {} or {} token after expression in argument list but found {} token.",
                        TokenKind::PuncComma,
                        TokenKind::PuncCloseParen,
                        self.current_token.kind
                    ));
                    break;
                }
            }
            let hi = self.record_current_pos();
            Expr {
                kind: ExprKind::FuncCall {
                    callee: identifier,
                    args,
                },
                span: lo.to(hi),
            }
        } else {
            let hi = self.record_current_pos();
            Expr {
                kind: ExprKind::Identifier(identifier),
                span: lo.to(hi),
            }
        }
    }

    fn parse_binary_expr_rhs(&mut self, min_bp: i8) -> Expr {
        // handle prefix operators
        let prefix_bp = self.current_token.kind.prefix_binding_power();
        let mut lhs = match prefix_bp {
            // not a prefix operator
            ((), -1) => self.parse_primary_expression(),
            // prefix operator
            _ => {
                let lo = self.record_current_pos();

                let ((), right_bp) = prefix_bp;
                let prefix_op = self.eat_token(); // eat prefix operator
                let rhs = self.parse_binary_expr_rhs(right_bp);

                let hi = self.record_current_pos();
                match prefix_op.kind {
                    TokenKind::OpPlus => rhs,
                    // build as 0 - x
                    TokenKind::OpMinus => Expr {
                        kind: ExprKind::BinaryExpr {
                            lhs: Box::new(Expr {
                                kind: ExprKind::Literal(0),
                                span: lo.zero_width_span(), // not an error but 0 does not actually appear in source.
                            }),
                            rhs: Box::new(rhs),
                            op_type: TokenKind::OpMinus,
                        },
                        span: lo.to(hi),
                    },
                    _ => unreachable!("Invalid prefix operator."),
                }
            }
        };

        loop {
            let (left_bp, right_bp) = self.current_token.kind.infix_binding_power();

            // stop parsing
            if left_bp < min_bp {
                break;
            }
            let binop = self.eat_token(); // eat bin op

            let rhs = self.parse_binary_expr_rhs(right_bp);

            /*
            Record positions for diagnostics
                  lhs              hi
            lhs.lo   lhs.hi   hi.lo  hi.hi
            */
            let lo = lhs.span.lo;
            let hi = rhs.span.hi;

            lhs = Expr {
                kind: ExprKind::BinaryExpr {
                    lhs: Box::new(lhs),
                    rhs: Box::new(rhs),
                    op_type: binop.kind,
                },
                span: lo.to(hi),
            }
        }

        lhs
    }

    fn parse_fn_prototype(&mut self) -> FnProto {
        self.expect_and_eat_tok(TokenKind::KeywordFn);
        let iden = self.expect_and_eat_iden_tok();
        self.expect_and_eat_tok(TokenKind::PuncOpenParen);

        // parse argument list
        let mut args: Vec<String> = Vec::new();
        loop {
            if self.current_token.kind == TokenKind::PuncCloseParen {
                self.expect_and_eat_tok(TokenKind::PuncCloseParen);
                break; // exit loop
            }
            // parse argument identifier
            let arg_iden = self.expect_and_eat_iden_tok();
            args.push(arg_iden);

            if self.current_token.kind == TokenKind::PuncComma {
                self.expect_and_eat_tok(TokenKind::PuncComma);
            } else if self.current_token.kind != TokenKind::PuncCloseParen {
                self.emit_err_at_current_tok(format!(
                    "Expected {} or {} token after identifier in argument list but found {} token.",
                    TokenKind::PuncComma,
                    TokenKind::PuncCloseParen,
                    self.current_token.kind
                ));
                break;
            }
        }

        FnProto { args, iden }
    }

    fn parse_fn_declaration(&mut self) -> Stmt {
        let lo = self.record_current_pos();
        let is_extern = {
            if self.current_token.kind == TokenKind::KeywordExtern {
                self.eat_token(); // eat 'extern' keyword
                true
            } else {
                false
            }
        };

        let proto = self.parse_fn_prototype();

        if !is_extern {
            let body = self.parse_block_statement();
            let hi = self.record_current_pos();
            Stmt {
                kind: StmtKind::Fn {
                    proto,
                    body: Some(body),
                },
                span: lo.to(hi),
            }
        } else {
            self.expect_and_eat_tok(TokenKind::PuncSemi);
            let hi = self.record_current_pos();
            Stmt {
                kind: StmtKind::Fn { proto, body: None },
                span: lo.to(hi),
            }
        }
    }

    fn parse_block_statement(&mut self) -> Block {
        self.expect_and_eat_tok(TokenKind::PuncOpenBrace);
        let mut stmts: Vec<Stmt> = Vec::new();
        loop {
            if self.current_token.kind == TokenKind::PuncCloseBrace
                || self.current_token.kind == TokenKind::EndOfFile
            // if TokenKind::EndOfFile, self.expect_and_eat_tok will create an error.
            {
                self.expect_and_eat_tok(TokenKind::PuncCloseBrace);
                break;
            }
            let statement = self.parse_statement();
            stmts.push(statement);
        }

        Block { stmts }
    }

    fn parse_while_stmt(&mut self) -> Stmt {
        let lo = self.record_current_pos();
        self.expect_and_eat_tok(TokenKind::KeywordWhile);
        let condition = self.parse_expression();
        let body = self.parse_block_statement();
        let hi = self.record_current_pos();

        Stmt {
            kind: StmtKind::While { condition, body },
            span: lo.to(hi),
        }
    }

    fn parse_if_else_stmt(&mut self) -> Stmt {
        let lo = self.record_current_pos();
        self.expect_and_eat_tok(TokenKind::KeywordIf);
        let if_condition = self.parse_expression();
        let if_body = self.parse_block_statement();

        let else_body = if self.current_token.kind == TokenKind::KeywordElse {
            self.expect_and_eat_tok(TokenKind::KeywordElse);
            let else_body = self.parse_block_statement();
            Some(else_body)
        } else {
            None
        };
        let hi = self.record_current_pos();

        Stmt {
            kind: StmtKind::IfElse {
                if_condition,
                if_body,
                else_body,
            },
            span: lo.to(hi),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    /// Utility function for tests
    fn parse_str_as_repl(s: &str) -> Stmt {
        let source = Rc::new(SourceFile::new("<test>", s));
        let mut lexer_error_reporter = ErrorReporter::new(source.clone());
        let mut lexer = Lexer::with_source_file(&source, &mut lexer_error_reporter);
        let mut parser_error_reporter = ErrorReporter::new(source.clone());
        let mut parser = Parser::new(&mut lexer, &mut parser_error_reporter);

        let res = parser.parse_repl_input();

        let mut errors = ErrorReporter::new(source.clone());
        errors.merge_from(&lexer_error_reporter);
        errors.merge_from(&parser_error_reporter);
        assert!(!errors.has_errors());
        res
    }

    /// Utility function for tests
    fn parse_str_as_expr(s: &str) -> Expr {
        let source = Rc::new(SourceFile::new("<test>", s));
        let mut lexer_error_reporter = ErrorReporter::new(source.clone());
        let mut lexer = Lexer::with_source_file(&source, &mut lexer_error_reporter);
        let mut parser_error_reporter = ErrorReporter::new(source.clone());
        let mut parser = Parser::new(&mut lexer, &mut parser_error_reporter);

        let res = parser.parse_expression();

        let mut errors = ErrorReporter::new(source.clone());
        errors.merge_from(&lexer_error_reporter);
        errors.merge_from(&parser_error_reporter);
        assert!(!errors.has_errors());
        res
    }

    /// Utility function for tests
    fn parse_str_as_stmt(s: &str) -> Stmt {
        let source = Rc::new(SourceFile::new("<test>", s));
        let mut lexer_error_reporter = ErrorReporter::new(source.clone());
        let mut lexer = Lexer::with_source_file(&source, &mut lexer_error_reporter);
        let mut parser_error_reporter = ErrorReporter::new(source.clone());
        let mut parser = Parser::new(&mut lexer, &mut parser_error_reporter);

        let res = parser.parse_statement();

        let mut errors = ErrorReporter::new(source.clone());
        errors.merge_from(&lexer_error_reporter);
        errors.merge_from(&parser_error_reporter);
        assert!(!errors.has_errors());
        res
    }

    /// Utility function for tests
    fn parse_str_as_unit(s: &str) -> CompilationUnit {
        let source = Rc::new(SourceFile::new("<test>", s));
        let mut lexer_error_reporter = ErrorReporter::new(source.clone());
        let mut lexer = Lexer::with_source_file(&source, &mut lexer_error_reporter);
        let mut parser_error_reporter = ErrorReporter::new(source.clone());
        let mut parser = Parser::new(&mut lexer, &mut parser_error_reporter);

        let res = parser.parse_compilation_unit();

        let mut errors = ErrorReporter::new(source.clone());
        errors.merge_from(&lexer_error_reporter);
        errors.merge_from(&parser_error_reporter);
        assert!(!errors.has_errors());
        res
    }

    #[test]
    fn parse_basic_bin_op() {
        assert_debug_snapshot!(parse_str_as_expr("1 + 1"));
        assert_debug_snapshot!(parse_str_as_expr("a + b"));
    }

    #[test]
    fn parse_bad_bin_op() {
        parse_str_as_expr("1 ++ 1");
        parse_str_as_expr("1 -- 1");
    }

    #[test]
    fn parse_operator_precedence() {
        assert_debug_snapshot!(parse_str_as_expr("1 + 2 * 3"));
        assert_debug_snapshot!(parse_str_as_expr("1 * 2 + 3 * 4"));
        assert_debug_snapshot!(parse_str_as_expr("my.field"));
        assert_debug_snapshot!(parse_str_as_expr("my.nested.field")); // test right associativity

        assert_debug_snapshot!(parse_str_as_expr("foo = 1")); // assignment operator
        assert_debug_snapshot!(parse_str_as_expr("foo = foo + 1"));

        assert_debug_snapshot!(parse_str_as_expr("foo == 1"));
        assert_debug_snapshot!(parse_str_as_expr("foo < 1"));
        assert_debug_snapshot!(parse_str_as_expr("foo > 1"));
        assert_debug_snapshot!(parse_str_as_expr("foo <= 1"));
        assert_debug_snapshot!(parse_str_as_expr("foo >= 1"));
    }

    #[test]
    fn parse_unary_operator() {
        assert_debug_snapshot!(parse_str_as_expr("+1"));
        assert_debug_snapshot!(parse_str_as_expr("-1"));
        assert_debug_snapshot!(parse_str_as_expr("-1 + 2"));
        assert_debug_snapshot!(parse_str_as_expr("1 + (-2)"));
        assert_debug_snapshot!(parse_str_as_expr("1 + -2")); // also valid
        assert_debug_snapshot!(parse_str_as_expr("-my_var"));
    }

    #[test]
    fn parse_function_call() {
        assert_debug_snapshot!(parse_str_as_expr("my_function()"));
        assert_debug_snapshot!(parse_str_as_expr("my_function(10)"));
        assert_debug_snapshot!(parse_str_as_expr("my_function(10, a_variable, 2 + 2)"));
        assert_debug_snapshot!(parse_str_as_expr("func(nested_func())"));
        assert_debug_snapshot!(parse_str_as_expr("func(nested_func()) + 2"));
    }

    #[test]
    fn parse_let_statement() {
        assert_debug_snapshot!(parse_str_as_stmt("let x = 1;"));
        assert_debug_snapshot!(parse_str_as_stmt("let x = 1 + 1;"));
        assert_debug_snapshot!(parse_str_as_stmt("let x = func(2);"));
    }

    #[test]
    fn parse_if_else_statement() {
        assert_debug_snapshot!(parse_str_as_stmt(
            "if x {
                println(1);
                // do something
            }"
        )); // no else block
        assert_debug_snapshot!(parse_str_as_stmt(
            "if x {
                // do something
            } else {
                // do something else
            }"
        ));
    }

    #[test]
    #[should_panic]
    fn parse_alone_else_block() {
        parse_str_as_stmt(
            "// no if block
            else {
                // do something else
            }",
        );
    }

    #[test]
    fn parse_while_statement() {
        assert_debug_snapshot!(parse_str_as_stmt("while x < 10 { println(x); x = x + 1; }"));
        assert_debug_snapshot!(parse_str_as_stmt("while x < 10 { }")); // empty body
    }

    #[test]
    fn parse_fn_definition() {
        assert_debug_snapshot!(parse_str_as_unit(
            "fn test() {
                return 1;
            }"
        ));
        assert_debug_snapshot!(parse_str_as_unit(
            "fn test() {
                return 1 + 1;
            }"
        ));
        assert_debug_snapshot!(parse_str_as_unit(
            "fn recursion() {
                return recursion();
            }"
        ));
        assert_debug_snapshot!(parse_str_as_unit(
            "fn test() {
                let tmp = 1;
                return tmp;
            }"
        ));
        assert_debug_snapshot!(parse_str_as_unit(
            "fn test() {
                let i = 0;
                while i < 10 {
                    println(i);
                    i = i + 1;
                }
                return i;
            }"
        ));
        assert_debug_snapshot!(parse_str_as_unit(
            "fn comments() {
                // this is a comment
                // return 0;
                return 1;
            }"
        ));
    }

    #[test]
    #[should_panic]
    fn parse_fn_with_missing_paren() {
        parse_str_as_unit(
            "fn no_paren {
                return 0;
            }",
        );
    }

    #[test]
    fn parse_repl_input() {
        assert_debug_snapshot!(parse_str_as_repl("1 + 1;")); // accepts expressions (expression statement)
        assert_debug_snapshot!(parse_str_as_repl("1 + 1")); // semi colon is optional
        assert_debug_snapshot!(parse_str_as_repl("fn func() { return 1; }")); // function declarations / definition
        assert_debug_snapshot!(parse_str_as_repl("func()")); // function call

        assert_debug_snapshot!(parse_str_as_repl("")); // should be StmtKind::Noop
    }
}
