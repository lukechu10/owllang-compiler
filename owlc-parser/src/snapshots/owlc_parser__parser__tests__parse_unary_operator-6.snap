---
source: owlc-parser/src/parser.rs
expression: "parse_str_as_expr(\"-my_var\")"
---
Expr {
    kind: BinaryExpr {
        lhs: Expr {
            kind: Literal(
                0,
            ),
            span: Span {
                lo: BytePos(
                    0,
                ),
                hi: BytePos(
                    0,
                ),
            },
        },
        rhs: Expr {
            kind: Identifier(
                "my_var",
            ),
            span: Span {
                lo: BytePos(
                    1,
                ),
                hi: BytePos(
                    7,
                ),
            },
        },
        op_type: OpMinus,
    },
    span: Span {
        lo: BytePos(
            0,
        ),
        hi: BytePos(
            7,
        ),
    },
}
